pkg/receive/handler_test.go

```diff
@@ -54,155 +52,157 @@ func TestDetermineWriteErrorCause(t *testing.T) {
+			err: errors.Wrap(merrors.New(
 				errors.New("foo"),
-				errutil.MultiError([]error{
+				merrors.New(
 					errors.New("bar"),
 					errors.New("qux"),
-				}),
-			}), "baz"),
+				).Err(),
+			).Err(), "baz"),
 			threshold: 1,
-			exp:       errors.New("baz: 2 errors: foo; 2 errors: bar; qux"),
+			exp:       errors.New("baz: 3 errors: foo; bar; qux"),
 		},
 		{
 			name: "matching multierror",
-			err: errutil.MultiError([]error{
+			err: merrors.New(
 				storage.ErrOutOfOrderSample,
 				errors.New("foo"),
 				errors.New("bar"),
-			}),
+			).Err(),
 			threshold: 1,
 			exp:       errConflict,
 		},
 		{
 			name: "matching but below threshold multierror",
-			err: errutil.MultiError([]error{
+			err: merrors.New(
 				storage.ErrOutOfOrderSample,
 				errors.New("foo"),
 				errors.New("bar"),
-			}),
+			).Err(),
 			threshold: 2,
 			exp:       errors.New("3 errors: out of order sample; foo; bar"),
 		},
 		{
 			name: "matching multierror many",
-			err: errutil.MultiError([]error{
+			err: merrors.New(
 				storage.ErrOutOfOrderSample,
 				errConflict,
 				status.Error(codes.AlreadyExists, "conflict"),
 				errors.New("foo"),
 				errors.New("bar"),
-			}),
+			).Err(),
 			threshold: 1,
 			exp:       errConflict,
 		},
 		{
 			name: "matching multierror many, one above threshold",
-			err: errutil.MultiError([]error{
+			err: merrors.New(
 				storage.ErrOutOfOrderSample,
 				errConflict,
 				tsdb.ErrNotReady,
 				tsdb.ErrNotReady,
 				tsdb.ErrNotReady,
 				errors.New("foo"),
-			}),
+			).Err(),
 			threshold: 2,
 			exp:       errNotReady,
 		},
 		{
 			name: "matching multierror many, both above threshold, conflict have precedence",
-			err: errutil.MultiError([]error{
+			err: merrors.New(
 				storage.ErrOutOfOrderSample,
 				errConflict,
 				tsdb.ErrNotReady,
 				tsdb.ErrNotReady,
 				tsdb.ErrNotReady,
 				status.Error(codes.AlreadyExists, "conflict"),
 				errors.New("foo"),
-			}),
+			).Err(),
 			threshold: 2,
 			exp:       errConflict,
 		},
 		{
 			name: "nested matching multierror",
-			err: errors.Wrap(errors.Wrap(errutil.MultiError([]error{
+			err: errors.Wrap(errors.Wrap(merrors.New(
 				storage.ErrOutOfOrderSample,
 				errors.New("foo"),
 				errors.New("bar"),
-			}), "baz"), "qux"),
+			).Err(), "baz"), "qux"),
 			threshold: 1,
 			exp:       errConflict,
 		},
 		{
 			name: "deep nested matching multierror",
-			err: errors.Wrap(errutil.MultiError([]error{
-				errutil.MultiError([]error{
+			err: errors.Wrap(merrors.New(
+				merrors.New(
 					errors.New("qux"),
 					status.Error(codes.AlreadyExists, "conflict"),
 					status.Error(codes.AlreadyExists, "conflict"),
-				}),
+				).Err(),
 				errors.New("foo"),
 				errors.New("bar"),
-			}), "baz"),
+			).Err(), "baz"),
 			threshold: 1,
-			exp:       errors.New("baz: 3 errors: 3 errors: qux; rpc error: code = AlreadyExists desc = conflict; rpc error: code = AlreadyExists desc = conflict; foo; bar"),
+			exp:       errConflict,
```

Comment:
Can you double check @kakkoyun ? I think this makes sense more actually 🤔

Reply:
AFAIR from previous comments, we didn't want to handle deeply nested errors like this. This was how I implemented while refactoring this. but @squat pointed me out that this is a behaviour change. I'm ok to have this change as long as we are aware of possible implecations.

Reply:
This is pretty subtle and can easily have big implications on replication. We need to spend a bit of time here analyzing the changes. It would be a lot easier to merge this if the commit maintained the same behavior and only changed the syntax to use the new errors package. My general approach is to keep PRs scoped to one purpose and avoid making other/unrelated changes wherever possible

Reply:
Yea, sure just was curious if that is not making sense more

Reply:
But what I gather... we don't know if that makes sense and we don't want to think about it now?

Reply:
Separare fix in this PR: https://github.com/thanos-io/thanos/pull/3833